compiler/rustc_infer/src/infer/base_struct.rs

```diff
@@ -0,0 +1,48 @@
+use crate::infer::at::{At, ToTrace};
+use crate::infer::sub::Sub;
+use crate::infer::{InferOk, InferResult};
+use rustc_middle::ty;
+use rustc_middle::ty::relate::{relate_generic_arg, RelateResult, TypeRelation};
+use rustc_middle::ty::{GenericArg, SubstsRef, Ty};
+use std::iter;
+
+pub fn base_struct<'a, 'tcx>(at: At<'a, 'tcx>, a: Ty<'tcx>, b: Ty<'tcx>) -> InferResult<'tcx, ()> {
+    let trace = ToTrace::to_trace(at.infcx.tcx, at.cause, true, a, b);
+    at.infcx.commit_if_ok(|_| {
+        let mut fields = at.infcx.combine_fields(trace, at.param_env, at.define_opaque_types);
+        let mut sub = Sub::new(&mut fields, true);
+        base_struct_tys(&mut sub, a, b)
+            .map(move |_| InferOk { value: (), obligations: fields.obligations })
+    })
+}
+
+pub fn base_struct_tys<'tcx>(
+    sub: &mut Sub<'_, '_, 'tcx>,
+    a: Ty<'tcx>,
+    b: Ty<'tcx>,
+) -> RelateResult<'tcx, ()> {
+    match (a.kind(), b.kind()) {
+        (&ty::Adt(a_def, a_substs), &ty::Adt(b_def, b_substs)) if a_def == b_def => {
+            base_struct_substs(sub, a, a_substs, b_substs)?;
+            Ok(())
+        }
+        _ => bug!("not adt ty: {:?} and {:?}", a, b),
+    }
+}
+
+fn base_struct_substs<'tcx>(
+    sub: &mut Sub<'_, '_, 'tcx>,
+    adt_ty: Ty<'tcx>,
+    a_subst: SubstsRef<'tcx>,
+    b_subst: SubstsRef<'tcx>,
+) -> RelateResult<'tcx, ()> {
+    let tcx = sub.tcx();
+    let variances = tcx.variances_of(adt_ty.ty_adt_def().expect("not a adt ty!").did());
+
+    iter::zip(a_subst, b_subst).enumerate().for_each(|(i, (a, b))| {
+        let _arg: RelateResult<'tcx, GenericArg<'tcx>> =
```

Comment:
Why are you throwing this result away?

Reply:
The purpose of this code is to instantiate the uninstantiated `infer` generic created for base_expr in `check_expr_struct_fields` by calling the `relate_generic_arg` function. If `Err` is returned, the generic type has been instantiated by the type of `base_expr` itself and is not constrained by the created struct (constrained ones hav been checked in `check_expr_struct_fields`).

Reply:
Perhaps we can have a more accurate approach, such as maintaining a non-constrained generic indexes list for `base_expr`.

Reply:
Given that this is purely for diagnostics, this approach might be fine, but including what you wrote above as a comment would be welcome.

Reply:
> Given that this is purely for diagnostics

I'm confused by this comment. This isn't purely diagnostic, I thought? My understanding is that this function affects type inference -- notably, `relate_generic_arg` is also not an atomic operation, so even if it returns `Err`, it can still partially constrain inference variables inside of that.

I'm not convinced that this approach is completely sound. I would be comfortable with something more along the lines of:

> such as maintaining a non-constrained generic indexes list for base_expr.

At least in that case it's easier to reason about when we're allowed to unify unconstrained inference variables and when we expect to raise errors.